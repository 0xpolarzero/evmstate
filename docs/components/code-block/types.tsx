import type { VariantProps } from "class-variance-authority";
import type { JSX } from "react";

import { codeBlockActionsVariants, codeBlockContainerVariants } from "./styles.js";

// -----------------------------------------------------------------------------
// Variant props
// -----------------------------------------------------------------------------

type CodeBlockActionsVariantProps = VariantProps<typeof codeBlockActionsVariants>;

type CodeBlockVariantProps = VariantProps<typeof codeBlockContainerVariants>;

// -----------------------------------------------------------------------------
// Component props
// -----------------------------------------------------------------------------

export type CodeBlockActionsProps = CodeBlockActionsVariantProps & {
  code: string;
};

export type CodeBlockLanguage =
  | "javascript"
  | "js"
  | "typescript"
  | "ts"
  | "jsx"
  | "tsx"
  | "solidity"
  | "sol"
  | "python"
  | "py"
  | "bash"
  | "sh"
  | "diff"
  | "none";

export type CodeBlockProps = Omit<JSX.IntrinsicElements["pre"], "children"> &
  CodeBlockVariantProps & {
    fileName?: string;
    language?: CodeBlockLanguage;
    logo?: React.FC<JSX.IntrinsicElements["svg"]>;
    highlightLines?: number[];
    showLineNumbers?: boolean;
    breakLines?: boolean;
    collapsible?: boolean;
    defaultCollapsed?: boolean;
    containerProps?: JSX.IntrinsicElements["div"];
    children: string | Object | Array<string | Object>;
  };
