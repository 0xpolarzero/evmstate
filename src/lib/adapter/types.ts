import { Hex } from "tevm";
import { SolcStorageLayout, SolcStorageLayoutTypeBase, SolcStorageLayoutTypes } from "tevm/bundler/solc";

import { AbiType, AbiTypeToPrimitiveType } from "./schema";

/* -------------------------------------------------------------------------- */
/*                              TYPE HELPERS                                  */
/* -------------------------------------------------------------------------- */

/** Makes all properties of an object readonly deeply */
export type DeepReadonly<T> = T extends (infer R)[]
  ? ReadonlyArray<DeepReadonly<R>>
  : T extends Function
    ? T
    : T extends object
      ? { readonly [K in keyof T]: DeepReadonly<T[K]> }
      : T;

/** Extract the final value type from a Solidity type ID */
export type ParseSolidityType<TypeId extends string, Types extends SolcStorageLayoutTypes> = TypeId extends keyof Types
  ? Types[TypeId] extends { label: infer Label extends string }
    ? Label
    : never
  : TypeId;

/* -------------------------------------------------------------------------- */
/*                           TYPE EXTRACTION UTILITIES                        */
/* -------------------------------------------------------------------------- */

/** Extract key type from a mapping declaration */
export type ExtractMappingKeyType<T extends string> = T extends `mapping(${infer KeyType} => ${string})`
  ? KeyType
  : never;

/** Extract value type from a mapping declaration */
export type ExtractMappingValueType<T extends string> = T extends `mapping(${string} => ${infer ValueType})`
  ? ValueType
  : never;

/** Extract base type from an array declaration */
export type ExtractArrayBaseType<T extends string> = T extends `${infer BaseType}[]` | `${infer BaseType}[${string}]`
  ? BaseType
  : never;

/** Get the Solidity key type and convert to appropriate TS type */
export type SolidityKeyToTsType<KeyType extends string, Types extends SolcStorageLayoutTypes> = KeyType extends AbiType
  ? AbiTypeToPrimitiveType<KeyType>
  : KeyType extends `t_${infer SolidityType}`
    ? SolidityType extends AbiType
      ? AbiTypeToPrimitiveType<SolidityType>
      : any
    : any;

/* -------------------------------------------------------------------------- */
/*                           STORAGE DATA TYPE MAPPING                        */
/* -------------------------------------------------------------------------- */

/** Map Solidity types to TypeScript return types */
export type SolidityTypeToTsType<T extends string, Types extends SolcStorageLayoutTypes> =
  // Handle primitive types
  T extends AbiType
    ? AbiTypeToPrimitiveType<T>
    : // Handle struct types
      T extends `struct ${string}`
      ? StructToObject<T, Types>
      : // Handle arrays
        T extends `${string}[]` | `${string}[${string}]`
        ? SolidityTypeToTsType<ExtractArrayBaseType<T>, Types>[]
        : // Handle mappings (return value type)
          T extends `mapping(${string} => ${string})`
          ? SolidityTypeToTsType<ExtractMappingValueType<T>, Types>
          : // Handle bytes/string
            T extends "bytes" | "string"
            ? string
            : // Default case
              unknown;

/** Convert a struct type to an object with fields */
export type StructToObject<StructName extends string, Types extends Record<string, any>> = {
  // TODO: replace with SolcStorageLayoutTypes and fix label index
  [TypeId in keyof Types]: Types[TypeId]["label"] extends StructName
    ? Types[TypeId] extends { members: readonly any[] }
      ? {
          [Member in Types[TypeId]["members"][number] as Member["label"]]: SolidityTypeToTsType<
            ParseSolidityType<Member["type"], Types>,
            Types
          >;
        }
      : never
    : never;
}[keyof Types];

/* -------------------------------------------------------------------------- */
/*                            MAPPING TYPE HELPERS                            */
/* -------------------------------------------------------------------------- */

/** Extract mapping key types with their corresponding TypeScript types */
export type GetMappingKeyTypePairs<
  T extends string,
  Types extends SolcStorageLayoutTypes,
  Result extends readonly [string, any][] = [],
> = T extends `mapping(${infer KeyType} => ${infer ValueType})`
  ? ValueType extends `mapping(${string} => ${string})`
    ? GetMappingKeyTypePairs<ValueType, Types, [...Result, [KeyType, SolidityKeyToTsType<KeyType, Types>]]>
    : [...Result, [KeyType, SolidityKeyToTsType<KeyType, Types>]]
  : Result;

/** Get just the Solidity type strings for mapping keys as a tuple */
export type GetMappingKeyTypes<T extends string, Types extends SolcStorageLayoutTypes> =
  GetMappingKeyTypePairs<T, Types> extends readonly [...infer Pairs]
    ? { [K in keyof Pairs]: Pairs[K] extends [infer SolType, any] ? SolType : never }
    : [];

/** Get just the TypeScript types for mapping keys as a tuple */
export type GetMappingKeyTsTypes<T extends string, Types extends SolcStorageLayoutTypes> =
  GetMappingKeyTypePairs<T, Types> extends readonly [...infer Pairs]
    ? { [K in keyof Pairs]: Pairs[K] extends [string, infer TsType] ? TsType : never }
    : [];

/**
 * Create a tuple type of mapping keys with their types (for display/debugging) Each element has both type and value
 * properties
 */
export type GetMappingKeysTuple<T extends string, Types extends SolcStorageLayoutTypes> =
  GetMappingKeyTypePairs<T, Types> extends readonly [...infer Pairs]
    ? {
        [K in keyof Pairs]: Pairs[K] extends [infer SolType, infer TsType] ? { type: SolType; value: TsType } : never;
      }
    : [];

/* -------------------------------------------------------------------------- */
/*                                 PARAMETERS                                 */
/* -------------------------------------------------------------------------- */

/** Parameters for array getData operations */
type ArrayDataParams =
  | { index: number } // Get single item
  | { startIndex: number; endIndex: number } // Get range
  | { indexes: number[] }; // Get multiple items

/** Parameters for getData based on type */
export type GetDataParams<T extends string, Types extends SolcStorageLayoutTypes> =
  // Mappings with typed keys
  T extends `mapping(${string} => ${string})`
    ? { keys: GetMappingKeyTsTypes<T, Types> }
    : // Arrays with flexible access patterns
      T extends `${string}[]` | `${string}[${string}]`
      ? ArrayDataParams
      : // Default (empty params)
        Record<string, never>;

/** Parameters for getSlot based on type */
export type GetSlotParams<T extends string, Types extends SolcStorageLayoutTypes> =
  // For mappings with typed keys
  T extends `mapping(${string} => ${string})`
    ? { keys: GetMappingKeyTsTypes<T, Types> }
    : // Arrays need index
      T extends `${string}[]` | `${string}[${string}]`
      ? { index: number }
      : // Default (empty params)
        Record<string, never>;

/* -------------------------------------------------------------------------- */
/*                               RETURN TYPES                                 */
/* -------------------------------------------------------------------------- */

/** Get return type based on access pattern */
type ArrayDataReturnType<
  BaseType extends string,
  Types extends SolcStorageLayoutTypes,
  Params extends ArrayDataParams,
> = Params extends { index: number } ? SolidityTypeToTsType<BaseType, Types> : SolidityTypeToTsType<BaseType, Types>[];

/** Get return type for getData based on type and params */
export type GetDataReturnType<
  T extends string,
  Types extends SolcStorageLayoutTypes,
  Params extends GetDataParams<T, Types>,
> =
  // For arrays, return depends on access pattern
  T extends `${infer BaseType}[]` | `${infer BaseType}[${string}]`
    ? Params extends ArrayDataParams
      ? ArrayDataReturnType<BaseType, Types, Params>
      : never
    : // For other types, standard conversion
      SolidityTypeToTsType<T, Types>;

// TODO: type recursively for nested complex types (but no logic for that yet)
export type DecodedSnapshot<T extends string, Types extends SolcStorageLayoutTypes> = T extends
  | `${infer BaseType}[]`
  | `${infer BaseType}[${string}]`
  ? { index: number; value: SolidityTypeToTsType<BaseType, Types> }
  : T extends `struct ${string}`
    ? StructField<T, Types>
    : SolidityTypeToTsType<T, Types>;

type StructField<T extends string, Types extends SolcStorageLayoutTypes, K = keyof StructToObject<T, Types>> = {
  member: K;
  value: StructToObject<T, Types>[K];
};

/* -------------------------------------------------------------------------- */
/*                              STORAGE VARIABLE                              */
/* -------------------------------------------------------------------------- */

/** A storage adapter with type-safe methods - base interface */
// TODO: will probably not use these if we use the classes
// interface BaseStorageAdapter<T extends string, Types extends SolcStorageLayoutTypes> {
//   /** Type information about this storage variable */
//   type: SolidityKeyToTsType<T, Types>;

//   /** Encoding type of this storage variable */
//   encoding: SolcStorageLayoutTypeBase["encoding"];

//   /** Number of bytes used by this storage variable */
//   byteLength: number;

//   /** Optional offset of this storage variable within its slot */
//   offset?: number;

//   /**
//    * Get the slot for this variable or its nested elements
//    *
//    * - For static arrays: baseSlot + index
//    * - For dynamic arrays: keccak256(baseSlot) + index
//    */
//   getSlot(params?: GetSlotParams<T, Types>): Hex;

//   /** Get the data stored at this location - asynchronous since it accesses storage */
//   getData<P extends GetDataParams<T, Types>>(params?: P): Promise<GetDataReturnType<T, Types, P>>;
// }

// /** Storage adapter interface for array types with additional methods */
// interface ArrayStorageAdapter<T extends string, Types extends SolcStorageLayoutTypes>
//   extends BaseStorageAdapter<T, Types> {
//   /** The base type of items in this array */
//   base: ExtractArrayBaseType<T>;

//   /** Get the length of the array */
//   getLength(): Promise<bigint>;
// }

// /** Storage adapter interface for mapping types with additional methods */
// interface MappingStorageAdapter<T extends string, Types extends SolcStorageLayoutTypes>
//   extends BaseStorageAdapter<T, Types> {
//   /** The Solidity types of the keys for this mapping */
//   keys: GetMappingKeyTypes<T, Types>;
// }

// /** Storage adapter interface for struct types with additional methods */
// interface StructStorageAdapter<T extends string, Types extends SolcStorageLayoutTypes>
//   extends BaseStorageAdapter<T, Types> {
//   /** The Solidity types of the fields for this struct */
//   fields: StructToObject<T, Types>;
// }

/** Storage variable with conditional array methods */
// export type StorageAdapter<T extends string, Types extends SolcStorageLayoutTypes> = T extends
//   | `${string}[]`
//   | `${string}[${string}]`
//   ? ArrayStorageAdapter<T, Types>
//   : T extends `mapping(${string} => ${string})`
//     ? MappingStorageAdapter<T, Types>
//     : T extends `struct ${string}`
//       ? StructStorageAdapter<T, Types>
//       : BaseStorageAdapter<T, Types>;
