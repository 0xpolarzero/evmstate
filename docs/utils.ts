import { isHex } from "tevm";
import clsx, { type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

const maxBytesLength = 64;
export const stringify = (value: any, space: number = 2) => {
  return JSON.stringify(
    value,
    (_, v) => {
      if (typeof v === "bigint") return `__bigint__${v.toString()}`;
      if (isHex(v) && (v.length - 2) / 2 > maxBytesLength)
        return `__long_hex__${v.slice(0, maxBytesLength + 2)}...${v.slice(-maxBytesLength)}`;
      return v;
    },
    space,
  );
};
