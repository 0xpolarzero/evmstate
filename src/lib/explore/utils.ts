import { Hex, toHex } from "tevm";
import { ByteArray, ToHexParameters } from "viem";

export const max = <T extends bigint | number>(...nums: Array<T>): T =>
  // @ts-expect-error bigint/number
  nums.reduce((max, num) => (num > max ? num : max), -Infinity);

/**
 * Converts a value to a hex string and ensures it has an even number of characters (full bytes)
 *
 * @param value The value to convert to hex
 * @param options Optional configuration options passed to tevm's toHex
 * @returns A hex string with full bytes
 */
export const toHexFullBytes = (value: string | number | bigint | boolean | ByteArray, opts?: ToHexParameters): Hex => {
  // Use tevm's toHex function
  const hex = toHex(value, opts);

  // If the hex string has an odd number of characters after the 0x prefix,
  // pad it with a leading zero to make it even
  if ((hex.length - 2) % 2 !== 0) return ("0x0" + hex.slice(2)) as Hex;
  return hex;
};
