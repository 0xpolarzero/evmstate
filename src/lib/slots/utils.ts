import { Address, decodeAbiParameters, encodeAbiParameters, Hex, isHex, toHex } from "tevm";
import { abi } from "@shazow/whatsabi";
import { AbiType, AbiTypeToPrimitiveType } from "abitype";
import { padHex } from "viem";

import { MappingKey } from "@/lib/slots/types";

/* -------------------------------------------------------------------------- */
/*                                  MAPPINGS                                  */
/* -------------------------------------------------------------------------- */

/** Extract values from a transaction trace that might be used as mapping keys */
// TODO: refactor to navigate the stack trace and identify both potential keys and array indexes
/**
 * Typical access patterns:
 *
 * Mapping
 *
 * - KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key, uint256(mappingSlot)))
 * - -> key is the input to the hash so we might even be able to retrieve the slot and compare directly
 * - KECCAK256 → KECCAK256 → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(abi.encode(key2, keccak256(abi.encode(key1, uint256(mappingSlot)))))
 * - -> same here we get each key from each keccak input
 *
 * Array
 *
 * - KECCAK256 → ADD → SLOAD/SSTORE
 * - -> Slot calculation: keccak256(uint256(arraySlot)) + index
 *
 * As reference:
 *
 * - Struct in array: KECCAK256 → ADD → ADD → SLOAD/SSTORE Slot calculation: (keccak256(uint256(arraySlot)) + index) +
 *   memberOffset
 * - Mapping to Struct with Mapping: KECCAK256 → ADD → KECCAK256 → SLOAD/SSTORE Slot calculation:
 *   keccak256(abi.encode(key2, (keccak256(abi.encode(key1, uint256(mappingSlot))) + memberOffset)))
 */
export const extractPotentialKeys = (
  trace: {
    uniqueStackValues?: Array<string>;
    relevantOps?: Array<{
      op: string;
      stack: Array<string>;
    }>;
  },
  addresses: Array<Address>,
  abiFunctions: Array<abi.ABIFunction>,
  txData?: Hex,
): MappingKey[] => {
  const keys: MappingKey[] = [];

  // Add touched addresses
  addresses.forEach((address) => {
    keys.push({
      hex: padHex(address, { size: 32 }),
      decoded: address,
      type: "address",
    });
  });

  // Extract parameters from transaction data
  if (txData && txData.length > 10) {
    const selector = txData.slice(0, 10);
    const inputs = abiFunctions.find((fn) => fn.selector === selector)?.inputs;

    if (inputs) {
      // Decode function inputs
      const params = decodeAbiParameters(inputs, `0x${txData.slice(10)}`);

      params.forEach((param, index) => {
        // If it's an array, add each element as a key
        if (Array.isArray(param)) {
          param.forEach((p) => {
            const type = inputs[index].type.replace("[]", "") as AbiType;

            if (type && type !== "string" && type !== "bytes") {
              keys.push({
                hex: padHex(encodeAbiParameters([{ type }], [p]), { size: 32 }),
                decoded: p as AbiTypeToPrimitiveType<typeof type>,
                type,
              });
            }
          });
        } else {
          // Otherwise just add the key straight up
          const type = inputs[index].type as AbiType;

          if (type && type !== "string" && type !== "bytes") {
            keys.push({
              hex: padHex(encodeAbiParameters([{ type }], [param]), { size: 32 }),
              decoded: param as AbiTypeToPrimitiveType<typeof type>,
              type,
            });
          }
        }
      });
    }
  }

  // Process stack values from the trace
  if (trace.uniqueStackValues?.length) {
    // Process unique stack values directly
    for (const stackValue of trace.uniqueStackValues) {
      keys.push({
        hex: isHex(stackValue) ? padHex(stackValue, { size: 32 }) : toHex(stackValue, { size: 32 }),
        type: undefined,
      });
    }
  }

  // Deduplicate keys
  const uniqueMap = new Map();
  // Add the new key only if it's not already in the map (and don't replace a key with a defined type)
  keys.forEach((k) => {
    if (!uniqueMap.has(k.hex) || k.type) uniqueMap.set(k.hex, k);
  });

  return Array.from(uniqueMap.values()).sort((a, b) => {
    // prefer address as it's more likely to be a key
    if (a.type === "address") return -1;
    if (b.type === "address") return 1;
    // prefer defined types
    if (a.type === undefined) return 1;
    if (b.type === undefined) return -1;

    return 0;
  });
};

/* -------------------------------------------------------------------------- */
/*                                    UTILS                                   */
/* -------------------------------------------------------------------------- */

/** A helper function to clean up trace objects by removing undefined or zero values */
export const cleanTrace = (obj: any) => {
  const { current, next, note, ...rest } = obj;
  let trace = { ...rest };

  // Only include note if it exists
  if (note) trace.note = note;

  // Same for current and next
  trace.current = { hex: current.hex };
  if (current.decoded !== undefined) trace.current = { hex: current.hex, decoded: current.decoded };

  if (next && rest.modified) {
    trace.next = { hex: next.hex };
    if (next.decoded !== undefined) trace.next = { hex: next.hex, decoded: next.decoded };
  }

  return trace;
};
