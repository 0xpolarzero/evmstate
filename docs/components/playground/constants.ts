import { EthjsAddress } from "tevm/utils";
import type { ExtractAbiFunctions } from "abitype";

import * as LAYOUTS from "@test/codegen/layouts/index.js";
import * as CONTRACTS from "@test/contracts/index.js";

export const contract = CONTRACTS["Playground"].withAddress(
  EthjsAddress.fromString("0x987C2AF139EAEaBdF8D6d3d1723C1883bEa1f2AF").toString(),
);
export const layout = LAYOUTS["Playground"];
export const callerAddress = EthjsAddress.fromString("0xCa11e40000000000000000000000000000000000");
export const localStorageKey = "EVMSTATE_PLAYGROUND_STATE";

export const functionDescriptions: Record<ExtractAbiFunctions<typeof contract.abi, "nonpayable">["name"], string> = {
  addValue: "Adds to a dynamic array",
  addUser: "Adds a user struct to a mapping and array",
  toggleUserActive: "Toggles a boolean in a struct within a mapping",
  setBalance: "Updates a simple mapping value",
  setAllowance: "Updates a nested mapping value",
  addTransaction: "Adds to a dynamic array within a mapping",
  updateBasicValues: "Updates primitive types (uint256, bool)",
  updatePackedValues: "Updates packed storage variables (uint8, uint16, uint32, bool)",
  setStringAndBytes: "Updates string and bytes storage",
  setFixedValue: "Updates a fixed-size array element",
};
