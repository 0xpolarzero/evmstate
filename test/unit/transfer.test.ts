import { createAddress } from "tevm/address";
import { EthjsAccount, Hex, parseEther } from "tevm/utils";
import { describe, expect, it } from "vitest";

import { ACCOUNTS, CONTRACTS } from "@test/constants";
import { getClient } from "@test/utils";
import { traceStorageAccess } from "@/index";
import { IntrinsicsDiff } from "@/lib/trace/types";

const { NativeTransfer, ETHReceiver } = CONTRACTS;
const { caller, recipient } = ACCOUNTS;

/**
 * Native Transfers tests
 *
 * This test suite verifies:
 *
 * 1. Contract to contract native token transfers
 * 2. EOA to EOA native token transfers
 */
describe("Native Transfers", () => {
  it("should track native token transfers between contracts", async () => {
    const client = getClient();
    // Fund the sender contract
    const vm = await client.transport.tevm.getVm();
    await vm.stateManager.putAccount(
      createAddress(NativeTransfer.address),
      EthjsAccount.fromAccountData({
        balance: parseEther("10"),
        nonce: 0n,
      }),
    );
    const transferAmount = parseEther("1");

    // Get initial balances
    const initialSenderBalance = await client.getBalance({
      address: NativeTransfer.address,
    });
    const initialReceiverBalance = await client.getBalance({
      address: ETHReceiver.address,
    });

    // Transfer ETH from NativeTransfer to ETHReceiver
    const trace = await traceStorageAccess({
      client,
      from: caller.toString(),
      to: NativeTransfer.address,
      abi: NativeTransfer.abi,
      functionName: "transferEth",
      args: [ETHReceiver.address, transferAmount],
    });

    // There should be no storage changes
    expect(trace[NativeTransfer.address].storage).toEqual({});
    // Balance should be updated
    expect(trace[NativeTransfer.address].intrinsic).toEqual({
      balance: {
        current: initialSenderBalance,
        next: initialSenderBalance - transferAmount,
      },
      nonce: { current: 0n },
      codeHash: { current: expect.any(String) },
      deployedBytecode: { current: expect.any(String) },
      storageRoot: { current: expect.any(String) },
    } as const satisfies IntrinsicsDiff);

    // Same for the receiver
    expect(trace[ETHReceiver.address].storage).toEqual({});
    expect(trace[ETHReceiver.address].intrinsic).toEqual({
      balance: {
        current: initialReceiverBalance,
        next: initialReceiverBalance + transferAmount,
      },
      nonce: { current: 0n },
      codeHash: { current: expect.any(String) },
      deployedBytecode: { current: expect.any(String) },
      storageRoot: { current: expect.any(String) },
    } as const satisfies IntrinsicsDiff);
  });

  it("should track native token transfers between EOAs", async () => {
    const client = getClient();
    const transferAmount = parseEther("1");
    const senderAddress = caller.toString();
    const recipientAddress = recipient.toString();

    // Get initial balances
    const initialSenderBalance = await client.getBalance({
      address: senderAddress,
    });
    const initialReceiverBalance = await client.getBalance({
      address: recipientAddress,
    });

    // Simulate the call to get the gas cost
    const { amountSpent } = await client.tevmCall({
      from: senderAddress,
      to: recipientAddress,
      value: transferAmount,
    });

    // Transfer ETH from caller to recipient
    const trace = await traceStorageAccess({
      client,
      from: senderAddress,
      to: recipientAddress,
      value: transferAmount,
    });

    // There should be no storage changes (it's an EOA)
    expect(trace[senderAddress].storage).toEqual({});
    // Balance should be updated
    expect(trace[senderAddress].intrinsic).toEqual({
      balance: {
        current: initialSenderBalance,
        next: initialSenderBalance - transferAmount - (amountSpent ?? 0n),
      },
      nonce: { current: 0n, next: 1n },
      codeHash: { current: expect.any(String) },
      deployedBytecode: { current: expect.any(String) },
      storageRoot: { current: expect.any(String) },
    } as const satisfies IntrinsicsDiff);

    // Same for the recipient
    expect(trace[recipientAddress].storage).toEqual({});
    expect(trace[recipientAddress].intrinsic).toEqual({
      balance: {
        current: initialReceiverBalance,
        next: initialReceiverBalance + transferAmount,
      },
      nonce: { current: 0n },
      codeHash: { current: expect.any(String) },
      deployedBytecode: { current: expect.any(String) },
      storageRoot: { current: expect.any(String) },
    } as const satisfies IntrinsicsDiff);
  });
});
