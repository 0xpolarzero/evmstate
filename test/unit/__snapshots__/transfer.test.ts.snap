// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Native Transfers > should track native token transfers between EOAs 1`] = `
{
  "0x0000000000000000000000000000000000000000": {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
  "0x0000000000000000000000000000000000000001": {
    "balance": {
      "current": 2000000000000000000n,
      "modified": true,
      "next": 999999999999852580n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": true,
      "next": 1,
    },
    "storage": {},
  },
  "0x0000000000000000000000000000000000000002": {
    "balance": {
      "current": 10000000000000000000n,
      "modified": true,
      "next": 11000000000000000000n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
}
`;

exports[`Native Transfers > should track native token transfers between contracts 1`] = `
{
  "0x0000000000000000000000000000000000000000": {
    "balance": {
      "current": 1000000000000000000000n,
      "modified": false,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
  "0x0000000000000000000000000000000000000001": {
    "balance": {
      "current": 10000000000000000000n,
      "modified": true,
      "next": 9999999999999778597n,
    },
    "code": {
      "current": "0x",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": true,
      "next": 1,
    },
    "storage": {},
  },
  "0x4554485265636569766572000000000000000000": {
    "balance": {
      "current": 0n,
      "modified": true,
      "next": 1000000000000000000n,
    },
    "code": {
      "current": "0x608060405260043610601e575f3560e01c806312065fe0146028576024565b36602457005b5f5ffd5b3480156032575f5ffd5b506039604d565b60405160449190606a565b60405180910390f35b5f47905090565b5f819050919050565b6064816054565b82525050565b5f602082019050607b5f830184605d565b9291505056fe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
  "0x4e61746976655472616e73666572000000000000": {
    "balance": {
      "current": 1000000000000000000n,
      "modified": true,
      "next": 0n,
    },
    "code": {
      "current": "0x60806040526004361061002c575f3560e01c806312065fe014610037578063e9bb84c21461006157610033565b3661003357005b5f5ffd5b348015610042575f5ffd5b5061004b610089565b6040516100589190610133565b60405180910390f35b34801561006c575f5ffd5b50610087600480360381019061008291906101d4565b610090565b005b5f47905090565b804710156100d3576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016100ca9061026c565b60405180910390fd5b8173ffffffffffffffffffffffffffffffffffffffff166108fc8290811502906040515f60405180830381858888f19350505050158015610116573d5f5f3e3d5ffd5b505050565b5f819050919050565b61012d8161011b565b82525050565b5f6020820190506101465f830184610124565b92915050565b5f5ffd5b5f73ffffffffffffffffffffffffffffffffffffffff82169050919050565b5f61017982610150565b9050919050565b6101898161016f565b8114610193575f5ffd5b50565b5f813590506101a481610180565b92915050565b6101b38161011b565b81146101bd575f5ffd5b50565b5f813590506101ce816101aa565b92915050565b5f5f604083850312156101ea576101e961014c565b5b5f6101f785828601610196565b9250506020610208858286016101c0565b9150509250929050565b5f82825260208201905092915050565b7f496e73756666696369656e742062616c616e63650000000000000000000000005f82015250565b5f610256601483610212565b915061026182610222565b602082019050919050565b5f6020820190508181035f8301526102838161024a565b905091905056fe0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "modified": false,
    },
    "nonce": {
      "current": 0,
      "modified": false,
    },
    "storage": {},
  },
}
`;
